flink-core/src/main/java/org/apache/flink/api/common/state/StateDescriptor.java

```diff
@@ -343,6 +351,9 @@ private void writeObject(final ObjectOutputStream out) throws IOException {
 			// we have a default value
 			out.writeBoolean(true);
 
+			TypeSerializer<T> serializer = serializerAtomicReference.get();
+			checkNotNull(serializer, "Serializer not yet initialized.");
```

Comment:
IMO, would be cleaner to move this check to the beginning of else-branch.
Moving early-returns / validations as early as possible would improve readability.